Add blog post about securing wildfly applications on OpenShift with oidc

theashiot opened the pull request:
Depends on updates to the simple-webapp-oidc example: 
https://github.com/wildfly-security-incubator/elytron-examples/pull/184

The guide can be previewed at: 
https://theashiot.github.io/wildfly-elytron/blog/securing-wildfly-apps-oidc-openshift/

Commit:
Add blog post about securing wildfly applications on OpenShift with oidc

fjuma commented on _posts/2023-10-03-securing-wildfly-apps-oidc-openshift.adoc:
```diff
@@ -0,0 +1,346 @@
+
+For more information, see the Keycloak documentation on how to https://www.keycloak.org/docs/latest/server_admin/index.html#_oidc_clients[Managing OpenID Connect clients].
+
+. Click *Action*, *Download adapter config*, and select `Keycloak OIDC JSON` as the *Format Option* to see the connection parameters.
```
This appears as item 1) in a list on its own. Is that intended?

theashiot replied:
Removed this line because we won't update the oidc.json

fjuma commented on _posts/2023-10-03-securing-wildfly-apps-oidc-openshift.adoc:
```diff
@@ -0,0 +1,346 @@
+}
+----
+<1> Change to the value of `client-id` defined in Keycloak: `simple-webapp-oidc`.
+<2> Change to the value of `provider-url` to the Keycloak realm URL. Use this command to obtain the URL:
```
Instead of needing to update this line, I think we could also specify this using the OIDC_PROVIDER_URL environment variable in the Helm Chart configuration. 

We could potentially also update the elytron example to use "simple-webapp-oidc" for the client name instead of "myclient".

With those changes, we wouldn't actually need the user to make any updates to the oidc.json file.

WDYT?

theashiot replied:
I changed the client name in the blog to myclient, just in case the example is used elsewhere.

Used environment variable to specify keycloak URL. Therefore, i've removed the entire section of configuring the application.

fjuma commented on _posts/2023-10-03-securing-wildfly-apps-oidc-openshift.adoc:
```diff
@@ -0,0 +1,346 @@
+  contextDir: simple-webapp-oidc
+  ref: YOUR_BRANCH                                                  <2>
+deploy:
+  replicas: 1
```
I think this is where we could potentially add an env section, e.g.,

```
deploy:
     env:
        -name: OIDC_PROVIDER_URL
        -value: 
```

theashiot replied:
thanks! done

fjuma commented on _posts/2023-10-03-securing-wildfly-apps-oidc-openshift.adoc:
```diff
@@ -0,0 +1,346 @@
+</profiles>
+----
+<1> *wildfly-maven-plugin* provisions a WildFly server with the specified layers with our application deployed.
+<2> *elytron-oidc-client* automatically adds the native OIDC client to our WildFly installation.
```
s/OIDC client/OIDC client subsystem

theashiot replied:
fixed

fjuma commented:
@theashiot This looks great! I've just added some small comments.

theashiot commented:
Thanks for the review and suggestions @fjuma! Based on your suggestion of not updating oidc config, i've simplified the procedure: users will only need to clone the example application repo and update helm.yaml.

Because users don't configure oidc.json, i've added some explanation about it in the **Behind the Scenes** section. 

Lemme know what you think!

best,
ashwin

fjuma commented on _posts/2023-10-03-securing-wildfly-apps-oidc-openshift.adoc:
```diff
@@ -0,0 +1,360 @@
+    - name: OIDC_PROVIDER_URL
+      value: <KEYCLOAK_URL>    <1>
+----
+<1> Replace _KEYCLOAK_URL_ with the Keycloak URL obtained in the previous command.
```
s/_KEYCLOAK_URL_/<KEYCLOAK_URL>

theashiot replied:
done

fjuma commented on _posts/2023-10-03-securing-wildfly-apps-oidc-openshift.adoc:
```diff
@@ -0,0 +1,360 @@
+
+* Examine the  https://github.com/wildfly-security/elytron-examples/blob/main/simple-webapp-oidc/pom.xml[pom.xml] file.
++
+Notice that it contains an *openshift* profile.
```
Do you think we should add a sentence about what the openshift profile means?

theashiot replied:
Added:

"A profile in Maven lets you create a set of configuration values to customize your application build for different environments. The *openshift* profile in this example defines configurations required for OpenShift deployment."

WDYT?

fjuma commented on _posts/2023-10-03-securing-wildfly-apps-oidc-openshift.adoc:
```diff
@@ -0,0 +1,360 @@
+
+== Resources
+
+* https://docs.wildfly.org/29/Getting_Started_on_OpenShift.html[Getting Started with WildFly on OpenShift]
```
We should update the WildFly docs URLs to reference 30 instead of 29.

theashiot replied:
fixed

fjuma commented:
@theashiot Looks great, I've added some small comments.

@PrarthonaPaul Would you be able to try this guide out?

fjuma commented on _posts/2023-10-03-securing-wildfly-apps-oidc-openshift.adoc:
```diff
@@ -0,0 +1,360 @@
+----
+service/keycloak created
+route.route.openshift.io/keycloak created
+Warning: apps.openshift.io/v1 DeploymentConfig is deprecated in v4.14+, unavailable in v4.10000+
```
@PrarthonaPaul Would you be able to also check if you're seeing this warning too?

PrarthonaPaul replied:
Trying it out now!

PrarthonaPaul replied:
Prarthona Paul: It seems that for OpenShift 4.14 and later, they are deprecating deploymentConfig and are encouraging customers to use Deployment instead: https://access.redhat.com/articles/7041372
I see that we are using DeploymentConfig in [keycloak.yaml. ](https://raw.githubusercontent.com/keycloak/keycloak-quickstarts/latest/openshift/keycloak.yaml)But maybe we can use the Deployment object instead of that unless we need something specific that only DeploymentConfig offers? 
Developer sandbox uses OpenShift version 4.14.1. That's why the warning is there

fjuma replied:
That's great info, thanks Prarthona! Would you be able to create a Keycloak github issue here (if one doesn't already exist for this)?

https://github.com/keycloak/keycloak/issues

PrarthonaPaul replied:
Done!

fjuma replied:
Great, please add the link to the issue here so we have it for reference later on. Thanks!

PrarthonaPaul replied:
here you go: https://github.com/keycloak/keycloak/issues/24607

fjuma commented on _posts/2023-10-03-securing-wildfly-apps-oidc-openshift.adoc:
```diff
@@ -0,0 +1,360 @@
+
+* Examine the  https://github.com/wildfly-security/elytron-examples/blob/main/simple-webapp-oidc/pom.xml[pom.xml] file.
++
+Notice that it contains an *openshift* profile. A profile in Maven lets you create a set of configuration values to customize your application build for different environments. The *openshift* profile in this example defines configurations required for OpenShift deployment.
```
"configurations required for OpenShift deployment"

Maybe could reword to something like:

"configuration that will be used by the WildFly Helm Chart when provisioning the WildFly server on OpenShift"

WDYT?

theashiot replied:
yep, this explains it better, updated.

PrarthonaPaul commented on _posts/2023-10-03-securing-wildfly-apps-oidc-openshift.adoc:
```diff
@@ -0,0 +1,360 @@
+
+[source]
+----
+helm install oidc-app -f /PATH/TO/ELYTRON/EXAMPLES/simple-webapp-saml/charts/helm.yaml wildfly/wildfly
```
Would it be useful to add `helm repo add wildfly https://docs.wildfly.org/wildfly-charts/` before this? 
If this is their first time using the WildFly helm chart repo, they would not have it added and would get an error.

fjuma replied:
That's a good question. We do mention WildFly Helm Chart in the pre-requisites and link to this page where adding the repo is mentioned:

https://docs.wildfly.org/30/Getting_Started_on_OpenShift.html#helm-charts

It's also important to update the WildFly Helm Chart if you've already added it. We should probably create an issue for mentioning that on the helm-charts page.

fjuma replied:
@theashiot WDYT?

theashiot replied:
Just to make it easier for the users, how about rather than mentioning the prerequisite about helm chart, we add the required commands?

### Deploy the Example Application to WildFly on OpenShift

If you haven't already installed the WildFly helm chart, install it:
`helm repo add wildfly https://docs.wildfly.org/wildfly-charts/`

Update the helm chart to ensure you have the latest one:
`helm repo update`

We can deploy our example application to WildFly on OpenShift using the WildFly Helm Chart:

`helm install oidc-app -f /PATH/TO/ELYTRON/EXAMPLES/simple-webapp-saml/charts/helm.yaml wildfly/wildfly
`
...

fjuma replied:
Sounds good, Ashwin and thanks for the suggestion Prarthona!

"Update the helm chart to ensure you have the latest one:" -> I think that could be "If you've already installed the WildFly Helm Chart", be sure to update it to ensure you have the latest one:"

theashiot replied:
thanks, done. Also, moved WildFly helm chart to resources section and added helm installation to prerequisites.

fjuma commented on _posts/2023-10-03-securing-wildfly-apps-oidc-openshift.adoc:
```diff
@@ -0,0 +1,375 @@
+
+== Deploy the Example Application to WildFly on OpenShift
+
+If you haven't already installed the WildFly helm chart, install it:
```
s/helm/Helm

theashiot replied:
sorry, fixed

fjuma commented:
@theashiot Thanks for all the updates! Looks great!

Just added one final minor comment.

Please squash the commits and adjust the date in the .adoc file name and in the file itself and this is ready to be merged.

Thanks again!

theashiot commented:
Thanks, @fjuma, fixed the issues and squashed the commits.

best,

ashwin
